### PR TITLE
Make searchsorted a primitive

### PR DESCRIPTION
**Pull request by jakevdp**
Re-visiting of #9108 with a slightly smaller scope.

This change adds a new primitive `jax.lax.searchsorted_p`, as well as a new function `jax.lax.searchsorted()` that provides the implementation for the existing `jax.numpy.searchsorted`. The primitive is slightly more complicated than `jax.numpy.searchsorted` in that it handles an arbitrary number of leading batch dimensions, and an arbitrary search axis for multi-dimensional inputs. Importantly, these additions make it closed under batching (that is, `vmap(searchsorted)` is always equivalent to a single call to `searchsorted`).

My reasoning for this proposal is that binary search is a reasonably common operation that cannot be straightforwardly expressed in terms of existing primitive operations (we currently have three implementations available that users can choose from), and so it makes sense to be treating `searchsorted` as a distinct operation.

There are a few benefits to this:

- jaxprs using `searchsorted` are much clearer
- Because the primitive is closed under batching, the batching rule for `searchsorted` is now trivial, where previously the implementation had to be traced op-by-op by the `BatchTracer`.
- Because `searchsorted` returns an integer index, the autodiff rule for `searchsorted` is trivial, where previously the implementation had to be traced op-by-op by the `JVPTracer`.
- this would pave the way for various backends to use more efficient `searchsorted` implementations rather than being beholden to the particular implementation in `jax.numpy`.

Here's an example of a jaxpr with this PR:
```python
In [1]: import jax
   ...: import jax.numpy as jnp
   ...: x = jnp.arange(10)
   ...: y = jnp.array([2, 4, 7])
   ...: jax.make_jaxpr(jnp.searchsorted)(x, y)
Out[1]: 
{ lambda ; a:i32[10] b:i32[3]. let
    c:i32[3] = searchsorted[batch_dims=0 dimension=0 method=default side=left] a b
  in (c,) }
```
And here's the same thing on the main branch:
```python
In [1]: import jax
   ...: import jax.numpy as jnp
   ...: x = jnp.arange(10)
   ...: y = jnp.array([2, 4, 7])
   ...: jax.make_jaxpr(jnp.searchsorted)(x, y)
Out[1]: 
{ lambda ; a:i32[10] b:i32[3]. let
    c:i32[3] = pjit[
      jaxpr={ lambda ; d:i32[10] e:i32[3]. let
          f:i32[3] = broadcast_in_dim[broadcast_dimensions=() shape=(3,)] 0
          g:i32[3] = broadcast_in_dim[broadcast_dimensions=() shape=(3,)] 10
          _:i32[] _:i32[3] h:i32[3] = scan[
            jaxpr={ lambda ; i:i32[10] j:i32[3] k:i32[] l:i32[3] m:i32[3]. let
                n:i32[] = add k 1
                o:i32[3] = add l m
                p:i32[3] = pjit[
                  jaxpr={ lambda ; q:i32[3] r:i32[]. let
                      s:i32[] = convert_element_type[
                        new_dtype=int32
                        weak_type=False
                      ] r
                      t:i32[3] = div q s
                      u:i32[3] = sign q
                      v:i32[] = sign s
                      w:bool[3] = ne u v
                      x:i32[3] = rem q s
                      y:bool[3] = ne x 0
                      z:bool[3] = and w y
                      ba:i32[3] = sub t 1
                      bb:i32[3] = pjit[
                        jaxpr={ lambda ; bc:bool[3] bd:i32[3] be:i32[3]. let
                            bf:i32[3] = select_n bc be bd
                          in (bf,) }
                        name=_where
                      ] z ba t
                    in (bb,) }
                  name=floor_divide
                ] o 2
                bg:bool[3] = lt p 0
                bh:i32[3] = add p 10
                bi:i32[3] = select_n bg p bh
                bj:i32[3,1] = broadcast_in_dim[
                  broadcast_dimensions=(0,)
                  shape=(3, 1)
                ] bi
                bk:i32[3,1] = gather[
                  dimension_numbers=GatherDimensionNumbers(offset_dims=(1,), collapsed_slice_dims=(), start_index_map=(0,))
                  fill_value=None
                  indices_are_sorted=False
                  mode=GatherScatterMode.PROMISE_IN_BOUNDS
                  slice_sizes=(1,)
                  unique_indices=False
                ] i bj
                bl:i32[3] = squeeze[dimensions=(1,)] bk
                bm:bool[3] = le j bl
                bn:i32[3] = pjit[
                  jaxpr={ lambda ; bc:bool[3] bd:i32[3] be:i32[3]. let
                      bf:i32[3] = select_n bc be bd
                    in (bf,) }
                  name=_where
                ] bm l p
                bo:i32[3] = pjit[
                  jaxpr={ lambda ; bc:bool[3] bd:i32[3] be:i32[3]. let
                      bf:i32[3] = select_n bc be bd
                    in (bf,) }
                  name=_where
                ] bm p m
              in (n, bn, bo) }
            length=4
            linear=(False, False, False, False, False)
            num_carry=3
            num_consts=2
            reverse=False
            unroll=1
          ] d e 0 f g
        in (h,) }
      name=searchsorted
    ] a b
  in (c,) }
```

### CODE REVIEW

**Comment by jakevdp** [comment]
Closing, as there's not much appetite for adding primitives without an associated XLA/HLO lowering.
